platforms/android/library/src/main/java/io/element/android/wysiwyg/utils/HtmlToSpansParser.kt

```diff
@@ -46,464 +42,227 @@ internal class HtmlToSpansParser(
-                val end = spanToAdd.end.coerceIn(0, text.length)
-                text.setSpan(spanToAdd.span, start, end, spanToAdd.flags)
-            }
-            text.removePlaceholderSpans()
-            text.addAtRoomSpans()
+        val dom = Jsoup.parse(html)
+        val text = buildSpannedString {
+            val body = dom.body()
+            parseChildren(body)
+            addAtRoomSpans()
         }
         if (BuildConfig.DEBUG) text.assertOnlyAllowedSpans()
         return text
     }
 
-    override fun setDocumentLocator(locator: Locator?) {}
-
-    override fun startDocument() {}
+    private fun SpannableStringBuilder.parseChildren(element: Element) {
+        for (child in element.childNodes()) {
+            when (child) {
+                is Element -> parseElement(child)
+                is TextNode -> parseTextNode(child)
+            }
+        }
+    }
 
-    override fun endDocument() {}
+    private fun SpannableStringBuilder.parseElement(element: Element) {
+        when (element.tagName()) {
+            "a" -> parseLink(element)
+            "b", "strong" -> parseInlineFormatting(element, InlineFormat.Bold)
+            "i", "em" -> parseInlineFormatting(element, InlineFormat.Italic)
+            "u" -> parseInlineFormatting(element, InlineFormat.Underline)
+            "del" -> parseInlineFormatting(element, InlineFormat.StrikeThrough)
+            "code" -> parseInlineCode(element)
+            "ul", "ol" -> parseChildren(element)
+            "li" -> parseListItem(element)
+            "pre" -> parseCodeBlock(element)
+            "blockquote" -> parseQuote(element)
+            "p" -> parseParagraph(element)
+            "br" -> parseLineBreak(element)
+        }
+    }
 
-    override fun startPrefixMapping(prefix: String?, uri: String?) {}
+    // region: Handle parsing of tags into spans
 
-    override fun endPrefixMapping(prefix: String?) {}
+    private fun SpannableStringBuilder.parseTextNode(child: TextNode) {
+        val text = child.wholeText
+        if (text.isEmpty()) return
 
-    override fun startElement(uri: String?, localName: String, qName: String?, atts: Attributes?) {
-        handleStartTag(localName, atts)
+        val previousSibling = child.previousSibling() as? Element
+        if (previousSibling != null && previousSibling.isBlock) {
+            append('\n')
+        }
+        append(text)
     }
 
-    override fun endElement(uri: String?, localName: String, qName: String?) {
-        handleEndTag(localName)
+    private fun SpannableStringBuilder.parseInlineFormatting(element: Element, inlineFormat: InlineFormat) {
+        val span = when (inlineFormat) {
+            InlineFormat.Bold -> StyleSpan(Typeface.BOLD)
+            InlineFormat.Italic -> StyleSpan(Typeface.ITALIC)
+            InlineFormat.Underline -> UnderlineSpan()
+            InlineFormat.StrikeThrough -> StrikethroughSpan()
+            InlineFormat.InlineCode -> return
```

Comment:
This `return` statement is strange to me.

Reply:
I added some comments to clarify this.